### ISSUE TITLE
Configurable connection timeouts

### ISSUE THREAD

**Issue by rseichter** [rseichter/fangfrisch#29]
Connection timeouts should be configurable, at least at the data provider level and possibly at URL level. The current, hard coded timeout is not flexible enough to handle known slow connections. The new feature must take scheduling considerations into account, e.g. avoiding problems due to multiple Fangfrisch instances running concurrently.

**Comment by bsdice** [comment]
Brief update: I searched long and hard and aside from http://ftp.swin.edu.au/sanesecurity/ there is a different rsync mirror which exposes the files on http(s): https://mirror.rollernet.us/sanesecurity/

I modified the Arch Linux AUR package https://aur.archlinux.org/packages/python-fangfrisch like so:

```
pkgname=python-fangfrisch-custom
_name=fangfrisch
...
prepare() {
    cd "$_name-$pkgver"
    sed -i "s#http://ftp.swin.edu.au/sanesecurity/#https://mirror.rollernet.us/sanesecurity/#g" fangfrisch/config/sanesecurity.py
}'
```

**Comment by rseichter** [comment]
Thanks for the update. I do however wonder why you chose to patch a Python file instead of locally configuring a [user-defined provider](https://rseichter.github.io/fangfrisch/#_user_defined_providers)?

**Comment by bsdice** [comment]
Because the Sanesecurity directory is not small and I don't know what to include or exclude so I wanted just the defaults as-is.

**Comment by rseichter** [comment]
Fangfrisch default configuration settings are [well documented](https://rseichter.github.io/fangfrisch/#internalconf), and additionally, there is the `dumpconf` command. 🤓

**Comment by bsdice** [comment]
Thank you for sticking with me. I compared the dumpconf output. Any particular reason that

  url_ftm = ${prefix}sanesecurity.ftm
  url_white = ${prefix}sigwhitelist.ign2
  url_miscreant = ${prefix}MiscreantPunch099-Low.ldb

are not included? I left out spam.ldb and winnow.complex.patterns.ldb, because those seem old and orphaned. I only consider low+med false positive risk signatures. scamnailer.ndb is not listed anymore on https://sanesecurity.com/usage/signatures/

**Comment by rseichter** [comment]
Sanesecurity rule sets have apparently changed since Fangfrisch 1.6.1 was released. For example, I have not seen MiscreantPunch mentioned before, and `scamnailer.ndb` is indeed no longer listed (it is also disabled by default, so no harm done). Looks like it is time for a cleanup.

As for `spam.ldb` and `winnow.complex.patterns.ldb`: neither is a member of the default Fangfrisch URL set for Sanesecurity, so I don't understand what you mean by "leaving them out"? Does the Arch Linux package contain a custom configuration file?

**Comment by bsdice** [comment]
spam.ldb and winnow.complex.patterns.ldb is listed in the signatures page, also present on the mirror. But small and file date suggest orphaned. So I chose to leave them out. Arch delivers only a default file.

**Comment by rseichter** [comment]
Well, if you want to use Fangfrisch's defaults, you could create your own configuration file /path/to/fangfrisch.conf (I don't know where the customary location is for Arch Linux) and include the following segment:

```
[sanesecurity]
enabled = yes
prefix = https://mirror.rollernet.us/sanesecurity/
```

That should be all that is required, and you could also add more `url_mumble` lines. Run using `fangfrisch --conf /path/to/fangfrisch.conf ...`, of course.

**Comment by bsdice** [comment]
Arch has it in /etc/fangfrisch/fangfrisch.conf, also I have done just that, but renamed the section and added the missing URL mumbles and removed the !commented out high false positive risk entries. And a lowercase for filename_miscreant = miscreantpunch099-low.ldb.

**Comment by rseichter** [comment]
I published a [developer release](https://pypi.org/project/fangfrisch/1.7.0.dev1/) with a minimal implementation supporting a new `connection_timeout` configuration option. Version 1.7.0.dev1 does not come with any safeguards against too-large timeout values, so please be mindful if you want to test it. Updated documentation is available in the [branch's docs](https://github.com/rseichter/fangfrisch/tree/connect-timeout/docs) directory.

**Comment by rseichter** [comment]
Issue addressed in release 1.7.0.
